File: LeetCode/Group-字符串/345. 反转字符串中的元音字母 Reverse Vowels of a String/Solution.java
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

class Solution {
    public String reverseVowels(String s) {
        Set<Character> vowels = new HashSet<>(Arrays.asList('a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'));

        char[] charArray = s.toCharArray();
        int i = 0;
        int j = charArray.length - 1;

        while (i < j) {
            if (!vowels.contains(charArray[i]))
                i++;
            else if (!vowels.contains(charArray[j]))
                j--;
            else {
                char temp = charArray[i];
                charArray[i] = charArray[j];
                charArray[j] = temp;
                i++;
                j--;
            }
        }
        return new String(charArray);
        
    }
}

File: LeetCode/Group-字符串/345. 反转字符串中的元音字母 Reverse Vowels of a String/Solution.py
# 用双指针分别从左右两侧向中间遍历，便于同时找到两侧第n个元音字母来做交换
class Solution:
    def reverseVowels(self, s: str) -> str:
        l, r = 0, len(s) - 1
        chars = list(s)
        vowels = set('aeiouAEIOU')

        while l < r:
            if s[l] not in vowels:
                l += 1
            elif s[r] not in vowels:
                r -= 1
            else:
                chars[l], chars[r] = chars[r], chars[l]
                l, r = l + 1, r - 1
        return ''.join(chars)